c10/core/TensorTypeSet.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <c10/core/TensorTypeId.h>
+#include <c10/util/llvmMathExtras.h>
+#include <c10/util/Exception.h>
+#include <ostream>
+
+namespace c10 {
+
+// A representation of a set of TensorTypeIds.  A tensor may
+// have multiple tensor type ids, e.g., a Variable tensor
+// can also be a CPU tensor; the TensorTypeSet specifies what
+// type ids apply.  The internal representation is as a 64-bit
+// bit set (this means only 64 tensor type ids are supported).
+//
+// At the moment, there are no nontrivial uses of this set; tensors
+// are always singletons.  In the near future, this set will represent
+// variable? + tensor type id.  In the far future, it will be
+// variable? + profiling? + tracing? + lazy? + tensor type id.
+//
+// An undefined tensor is one with an empty tensor type set.
+class TensorTypeSet {
```

Comment:
bikeshedding: if this is a set of TensorTypeIds why is this not TensorTypeIdSet?

Disclaimer: I think "Type" should be banned in the same way "Backend" should be banned: people use the term to mean 10 different things, so we shouldn't name anything _actually_ that.

Reply:
Mostly because TensorTypeIdSet looks more ugly to me :)

Based on more recent design docs, we kind of want something like `TensorTypeId => DispatchKey` and then `TensorTypeSet => DispatchKeySet`. But that would have involved a lot of annoying renaming, which I wanted to avoid for the PRs.

Reply:
the end state sounds good to me though.